native/src/widget/text_input.rs

```diff
@@ -193,8 +209,38 @@ where
             .max_width(self.max_width)
             .height(Length::Units(text_size));
 
+        let text_visual_size = renderer.measure_value(
+            &self.value.to_string(),
+            text_size,
+            self.font,
+        );
+
         let mut text = layout::Node::new(limits.resolve(Size::ZERO));
-        text.move_to(Point::new(padding, padding));
+
+        let x = {
+            let width = limits.max().width;
+            let center = (width / 2.0) - (text_visual_size / 2.0);
+            let right = width - text_visual_size;
+            match self.horizontal_alignment {
+                HorizontalAlignment::Left => padding,
+                HorizontalAlignment::Center => {
+                    if center < padding {
+                        padding
+                    } else {
+                        center
+                    }
+                }
+                HorizontalAlignment::Right => {
+                    if right < padding {
+                        padding
+                    } else {
+                        right
+                    }
+                }
+            }
+        };
+
+        text.move_to(Point::new(x, padding));
```

Comment:
I am not convinced the `layout` logic needs to change to accomodate the alignment. Instead, I think we should change rendering and measuring functions.

Is there any particular reason you went with this approach?

Reply:
I initially tried with the `draw` logic itself. I was successful to align text and cursor, but Clip primitive was also bound to `text_bounds`.

I figured out that I was changing every x position for Text, Cursor Quad, and Clip, using kind of similar code every place for these measurements.

They all were dependent on the same `text_bounds` variable and I had to change the `x` position in `text_bounds`
for each of them, with similar kind of code (not exactly the same, thus the duplicity in code). Thus, I moved the logic to the layout call where `text_bounds` was initially created.

Even I am not sure if this is right, but while going through the code I figured out that `widget.draw` is basically called during the layout phase, i.e before every `RequestRedraw`, so thought this might be ok.

Reply:
So I tried implementing this on `draw` calls and have successfully implemented the cursor, cursor selection, and text alignments as per the `horizontal_alignment` set.

But now I am stuck with some wrong offset set for horizontal_alignment `Center` and `Right`.

![issue](https://user-images.githubusercontent.com/11786283/83548284-27e45280-a521-11ea-8af4-9122a1e9e94b.gif)

As you can see in the above gif, for Left alignment things work, but for Center/Right alignment the moment Clipping starts, it starts moving to the left, due to some wrong offset.

I tried to change the `measure_cursor_and_scroll_offset` functions call signature, but it is being used some other `offset` call, where I don't have access to `horizontal_alignment`.

Any suggestions for this situation would help me a lot. 🙇

Reply:
I fixed the above issue, but there are some other issues still.

I will try to figure it out before posting it here 👍